ampd/src/stellar/rpc_client.rs

```diff
@@ -16,31 +18,68 @@ pub enum Error {
     TxHash,
 }
 
+#[derive(Error, Debug)]
+pub enum TxParseError {
+    #[error("Invalid operation count for Soroban transaction: expected {expected}, got {actual}")]
+    InvalidOperationCount { expected: usize, actual: usize },
+    #[error("Unsupported transaction metadata version")]
+    UnsupportedMetadataVersion,
+}
+
 /// TxResponse parses XDR encoded TransactionMeta to ContractEvent type, and only contains necessary fields for verification
 #[derive(Debug)]
 pub struct TxResponse {
     pub transaction_hash: String,
     pub successful: bool,
-    pub contract_events: VecM<ContractEvent>,
+    pub contract_events: Vec<ContractEvent>,
 }
 
 const STATUS_SUCCESS: &str = "SUCCESS";
+const EXPECTED_SOROBAN_OPERATION_COUNT: usize = 1;
 
-impl From<(Hash, GetTransactionResponse)> for TxResponse {
-    fn from((transaction_hash, response): (Hash, GetTransactionResponse)) -> Self {
-        let contract_events = response
-            .result_meta
-            .and_then(|meta| match meta {
-                TransactionMeta::V3(data) => data.soroban_meta.map(|meta| meta.events),
-                _ => None,
-            })
-            .unwrap_or_default();
+impl TryFrom<(Hash, GetTransactionResponse)> for TxResponse {
+    type Error = TxParseError;
 
-        Self {
-            transaction_hash: transaction_hash.to_string(),
-            successful: response.status == STATUS_SUCCESS,
-            contract_events,
+    fn try_from(
+        (transaction_hash, response): (Hash, GetTransactionResponse),
+    ) -> Result<Self, Self::Error> {
+        let transaction_hash = transaction_hash.to_string();
+        if response.status != STATUS_SUCCESS {
+            return Ok(Self {
+                transaction_hash,
+                successful: false,
+                contract_events: vec![],
+            });
         }
+        let contract_events = match response.result_meta.as_ref() {
+            Some(TransactionMeta::V4(_)) => {
+                let contract_events = response.events.contract_events;
+                let operation_count = contract_events.len();
+
+                if operation_count != EXPECTED_SOROBAN_OPERATION_COUNT {
+                    return Err(TxParseError::InvalidOperationCount {
+                        expected: EXPECTED_SOROBAN_OPERATION_COUNT,
+                        actual: operation_count,
+                    });
+                }
+
+                contract_events.into_iter().next().unwrap_or_default()
```

Comment:
Using default is dangerous in case the logic is moved around/changes
```suggestion
                contract_events.get(0).expect("soroban operation not found")
```
Expect on it

Reply:
make sense. updated

Reply:
lint error: ```error: accessing first element with `contract_events.get(0)`
Error:   --> ampd/src/stellar/rpc_client.rs:67:17
   |
67 | /                 contract_events
68 | |                     .get(0)
   | |___________________________^ help: try: `contract_events.first()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#get_first
   = note: `-D clippy::get-first` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::get_first)]`
   ```
   thus, updated to use `first()`